packages/filesystem/src/browser/filesystem-preferences.ts

```diff
@@ -35,13 +36,14 @@ export const filesystemPreferenceSchema: PreferenceSchema = {
                 '**/.git/objects/**': true,
                 '**/.git/subtree-cache/**': true,
                 '**/node_modules/**': true
-            }
+            },
+            'scopes': PreferenceScope.Default | PreferenceScope.User | PreferenceScope.Workspace | PreferenceScope.Folders
```

Comment:
?

Reply:
What does VS Code do? I don't think magical numbers are not good for UI.

Reply:
@akosyakov are you asking what vsCode does to define the "scopes"? 
When I looked into the vsCode source code 2 months ago, I find vsCode used hard-coded rules, such as any preference whose names start with "window" shouldn't be applied at the folder level.

Reply:
vsCode has a registry that categorises all preferences into the followings:
```ts
export const allSettings: { properties: {}, patternProperties: {} } = { properties: {}, patternProperties: {} };
export const applicationSettings: { properties: {}, patternProperties: {} } = { properties: {}, patternProperties: {} };
export const windowSettings: { properties: {}, patternProperties: {} } = { properties: {}, patternProperties: {} };
export const resourceSettings: { properties: {}, patternProperties: {} } = { properties: {}, patternProperties: {} };
```
From there the rest of the application knows if a preference can / cannot be applied to some certain "scope". for example, anything registered in `windowSettings` cannot be applied at the folder scope. 

in the preference schema, vsCode defined something very similar to what i have in this PR
```ts
	private registerJSONConfiguration(configuration: IConfigurationNode) {
		function register(configuration: IConfigurationNode) {
			let properties = configuration.properties;
			if (properties) {
				for (const key in properties) {
					allSettings.properties[key] = properties[key];
					switch (properties[key].scope) {
						case ConfigurationScope.APPLICATION:
							applicationSettings.properties[key] = properties[key];
							break;
						case ConfigurationScope.WINDOW:
							windowSettings.properties[key] = properties[key];
							break;
						case ConfigurationScope.RESOURCE:
							resourceSettings.properties[key] = properties[key];
							break;
					}
				}
			}
			let subNodes = configuration.allOf;
			if (subNodes) {
				subNodes.forEach(register);
			}
		}
		register(configuration);
		this._onDidSchemaChange.fire(); 
	}
```
the reason that i used `1-2-4-8` is 
- having one number to represent all 4 scopes, and the list of scopes can be extended as long as the integer does not overflow
- giving developers who add / edit  preference schema more flexibility: with my code, i can create a preference with a scope of any combination of the 4 (i.e., Default, User, Workspace, Folder), e.g., something that makes sense at the `Workspace` scope and is invalid at the `User` scope. While in vsCode, i would have choose one the three scopes: [APPLICATION, WINDOW, RESOURCE].

Reply:
I think numbers are not good for end users. It should be readable and brainer to type.

// cc @svenefftinge

Reply:
How VS Code `Application`, `Window`, `Resource` scopes are mapped to our scopes?

Reply:
>I think numbers are not good for end users. It should be readable and no brainer to type.

I am using enums to avoid numbers showing up, so end users **don't** need to type any numbers. They only need to 
- not specify the `scopes` at all (in this case the `scopes` is Preference.Default + Preference.User + Preference.Workspace), or
- use any combinations of the 4 scopes that the `PreferenceScope` enum defines.

Reply:
> How VS Code Application, Window, Resource scopes are mapped to our scopes?

@akosyakov 
here is the mapping between scopes defined in vsCode and this PR:
vsCode.ConfigurationScope.APPLICATION => PreferenceScope.Default + PreferenceScope.User
vsCode.ConfigurationScope.WINDOW => PreferenceScope.Default + PreferenceScope.User + PreferenceScope.Workspace
vsCode.ConfigurationScope.RESOURCE => PreferenceScope.Default + PreferenceScope.User + PreferenceScope.Workspace + PreferenceScope.Folder

With the scopes defined in vsCode, developers are unable to define a preference that is, for example, only valid at the `Workspace` and `Default` level (if i use `vsCode.ConfigurationScope.WINDOW`, users can use it at the `User` level). While with code in this PR, developers have that flexibility.

Reply:
ok

introducing similar constants would be handy, i.e. `PreferenceScope.Resource` would be handy, instead of typing `PreferenceScope.Default | PreferenceScope.User | PreferenceScope.Workspace | PreferenceScope.Folders`

Reply:
sure. changed as suggested.

Reply:
We should stick to VS Code and pure JSON approach:
- otherwise, it is hard to move preferences from VS Code extensions to Theia extensions;
- If we move preferences to JSON file at some point, it won't be possible to use enums from there.

I should be called `scope` and has the same values as VS Code does.

// cc @svenefftinge

Reply:
OK, I will define two enums:
```ts
export const enum PreferenceScope { 
        // this is from vsCode, supposed to be used by pref schemas
	APPLICATION = 1,
	WINDOW,
	RESOURCE,
}

export const enum PreferenceProviderScope {
   // this is the enum I already have, supposed to be used by PrefProviders internally
   Default = 1,
   User = 2,
   WS = 4,
   Folder = 8
}
```
This way we would have the compatibility with vsCode and possibility of moving to JSON files. What do you think? @akosyakov

Reply:
You should be able to type in JSON file `'scope': 'window'` and it should work, no `'scope': 1` or `'scope': 'APPLICATION'`.

Do you have a real use case for `developers are unable to define a preference that is, for example, only valid at the Workspace and Default level`? If not, please, let's think about it when it exists.

Reply:
>You should be able to type in JSON file 'scope': 'window' and it should work, no 'scope': 1 or 'scope': 'APPLICATION'.

`APPLICATION = 1` is from the vsCode source code, not something I invented :) 
I personally don't think 'scope': 'window' or 'scope': 1 or 'scope': 'APPLICATION' in the json file would be a problem, we could make any (or all) of them work in future :)

>Do you have a real use case for developers are "unable to define a preference that is, for example, only valid at the Workspace and Default level"? If not, please, let's think about it when it exists.

just want to clarify: in the latest proposal described above, `PreferenceScope` and `PreferenceProviderScope` will both be defined - I wasn't promoting using one instead of the other. And only `PreferenceScope` is supposed to used in the preference schema definition. Therefore, defining `a preference that is, for example, only valid at the Workspace and Default level` is not possible any more. 

`PreferenceProviderScope` is only used internally as an enum to create PreferenceProviders. 

Speaking of `a preference that is, for example, only valid at the Workspace and Default level`, yes, I have something in mind but not sure when it will come into reality. Think about a pair programming extension that connects 2 or more developers working simultaneously on the same project (https://github.com/theia-ide/theia/issues/1405). Different developers, in many cases, prefer to have different settings, e.g., a pref that toggles the highlight of the line that the other developer is working on (google docs has the highlight = on by default). With that "hightlight_the_other_guy = on", developers have better feel of collaboration and would less likely to step on each other's toes when editing the same package or file. imo, to have "hightlight_the_other_guy = on", we would need to add a new scope, which changes the PreferenceProviderScope into

```ts
export const enum PreferenceProviderScope {
   // this is the enum I already have, supposed to be used by PrefProviders internally
   Default = 1,
   User = 2,
   WS = 4,
   Folder = 8,
   PerInstance = 16 // or called InMemory
}
```
and it does not make sense to have that "hightlight_the_other_guy = on" defined in a scope other than `PerInstance`.

Reply:
@akosyakov hi Anton, are you ok with my proposal described above (i.e., having PreferenceScope defined in vsCode to achieve the compatibility, and PreferenceProviderScope for provider creation) ? I prefer to get your opinion before making this change, as it would change a lot of lines. Thanks !

Reply:
@elaihau my concern is to be compatible with VS code because of reasons I've put [above](https://github.com/theia-ide/theia/pull/3247#discussion_r237105884). If it would be possible then go ahead and do what you think is necessary.

Reply:
@akosyakov i updated the values used in `PreferenceScope` to keep them aligned with those used in vsCode.